blocks/bricks/sequence_generators.py

```diff
@@ -664,14 +665,15 @@ class SoftmaxEmitter(AbstractEmitter, Initializable, Random):
 
     """
     def __init__(self, initial_output=0, **kwargs):
-        self.initial_output = initial_output
         super(SoftmaxEmitter, self).__init__(**kwargs)
+        self.initial_output = initial_output
+        self.softmax = WithExtraDims(Softmax().apply)
```

Comment:
I'm not sure I like this. Why do we create many bricks for conceptually the same element? Maybe it is better for SoftMax to take an `axis` argument along which it works, with the last one as the default?

On a sidenote: Softmax shouldn't inherit from Activation, because it doesn't work element-wise. It acts axis-wise.

I coded this a while ago, so to  save on work:
```python
class AxisSoftmax(Brick):
    """
    Applies the Softmax function along a given axis (by default the last one)
    """
    def __init__(self, axis=-1, **kwargs):
        super(AxisSoftmax, self).__init__(**kwargs)
        self.axis = axis

    @application(inputs=['input_'], outputs=['output'])
    def apply(self, input_):
        input_ = input_ - input_.max(axis=self.axis, keepdims=True)
        ex = tensor.exp(input_)
        return ex / ex.sum(axis=self.axis, keepdins=True)

    @application(inputs=['input_'], outputs=['output'])
    def log_probs(self, input_):
        input_ = input_ - input_.max(axis=self.axis, keepdims=True)
        return input_ - tensor.log(tensor.exp(input_).sum(axis=self.axis,
                                                          keepdims=True))

    @application
    def categorical_cross_entropy(self, application_call,
                                  targets, input_,
                                  target_masks=None):
        """Return computationally stable softmax cost.

        Parameters
        ----------
        targets : :class:`~tensor.TensorVariable`
            In the case of a matrix argument, each slice along
            axis represents one distribution. In the vector case, each
            element represents the position of the '1' in a one hot-vector.
        input_ : :class:`~tensor.TensorVariable`
            Each slice along axis represents energies of a distribution,
            that is pre-softmax values.
        target_masks : None or :class:`~tensor.TensorVariable`
            If not None, the target_masks indicate which elemts to include
            in the cost
        average : bool
            Whether to compute the mean cost, or return individual costs

        Returns
        -------
        cost : :class:`~tensor.TensorVariable`
            The sum of the cross entropy between targets and Softmax(input_).

        """
        log_probs = self.log_probs(input_)

        application_call.add_auxiliary_variable(
            log_probs.copy(), name='log_probs')

        if targets.ndim == log_probs.ndim - 1:
            # move the selected axis to be the last!
            if self.axis != -1 and self.axis != targets.ndim:
                axis_order = range(log_probs.ndim)
                axnum = axis_order.pop(self.axis)
                axis_order.append(axnum)
                _log_prob = log_probs.dimshuffle(*axis_order)
            else:
                _log_prob = log_probs
            _log_prob = _log_prob.reshape((_log_prob.shape[:-1].prod(),
                                           _log_prob.shape[-1]))
            _targets = targets.ravel()
            cost = -_log_prob[tensor.arange(_targets.shape[0]),
                              _targets]
            cost = cost.reshape(targets.shape)
        elif targets.ndim == log_probs.ndim:
            cost = -(log_probs * targets).sum(axis=self.axis)
        else:
            raise TypeError('rank mismatch between x and y')
        if target_masks:
            cost = cost * target_masks
        return cost.sum()

    @application(inputs=['input_'], outputs=['output'])
    def classify(self, input_):
        return input_.argmax(axis=self.axis)

    @application
    def misclasification_count(self, application_call,
                         targets, input_,
                         target_masks=None):
        """Return computationally stable softmax cost.

        Parameters
        ----------
        targets : :class:`~tensor.TensorVariable`
           Matrix of integers encoding the proper class in a 1-hot fashion
        input_ : :class:`~tensor.TensorVariable`
            Each slice along axis represents energies of a distribution,
            that is pre-softmax values.
        target_masks : None or :class:`~tensor.TensorVariable`
            If not None, the target_masks indicate which elemts to include in
            the cost
        average : bool
            Whether to compute the mean cost, or return individual costs

        Returns
        -------
        cost : :class:`~tensor.TensorVariable`
            The sum of the cross entropy between targets and Softmax(input_).

        """
        predictions = self.classify(input_)

        application_call.add_auxiliary_variable(
            predictions.copy(), name='predictions')

        if targets.ndim == predictions.ndim:
            errs = tensor.neq(targets, predictions)
        else:
            raise TypeError('rank mismatch between x and y')
        if target_masks:
            errs = errs * target_masks
        return errs.sum().astype(theano.config.floatX)
```